File: mobile/lib/services/websocket_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as status;

enum WSMessageType {
  // Client → Server
  play,
  pause,
  seek,
  skip,
  trackChanged,
  timeSyncResponse,
  startMirrorMode,
  stopMirrorMode,

  // Server → Client
  playCommand,
  pauseCommand,
  seekCommand,
  skipCommand,
  memberJoined,
  memberLeft,
  playlistUpdated,
  timeSyncRequest,
  timeSyncResult,
  roomState,
  mirrorModeStarted,
  mirrorModeStopped,
}

/// Maps Go-backend snake_case message types to enum values
WSMessageType? _parseMessageType(String typeStr) {
  switch (typeStr) {
    case 'play': return WSMessageType.playCommand;
    case 'pause': return WSMessageType.pauseCommand;
    case 'seek': return WSMessageType.seekCommand;
    case 'skip': return WSMessageType.skipCommand;
    case 'track_changed': return WSMessageType.trackChanged;
    case 'member_joined': return WSMessageType.memberJoined;
    case 'member_left': return WSMessageType.memberLeft;
    case 'playlist_updated': return WSMessageType.playlistUpdated;
    case 'time_sync_request': return WSMessageType.timeSyncRequest;
    case 'time_sync_response': return WSMessageType.timeSyncResponse;
    case 'time_sync_result': return WSMessageType.timeSyncResult;
    case 'room_state': return WSMessageType.roomState;
    case 'mirror_mode_started': return WSMessageType.mirrorModeStarted;
    case 'mirror_mode_stopped': return WSMessageType.mirrorModeStopped;
    default:
      print('Unknown message type: $typeStr');
      return null;
  }
}

/// Maps enum values to Go-backend snake_case strings for sending
String _toWireType(WSMessageType type) {
  switch (type) {
    case WSMessageType.play: return 'play';
    case WSMessageType.pause: return 'pause';
    case WSMessageType.seek: return 'seek';
    case WSMessageType.skip: return 'skip';
    case WSMessageType.trackChanged: return 'track_changed';
    case WSMessageType.timeSyncResponse: return 'time_sync_response';
    case WSMessageType.startMirrorMode: return 'start_mirror_mode';
    case WSMessageType.stopMirrorMode: return 'stop_mirror_mode';
    default: return type.name;
  }
}

class WebSocketService {
  final String baseUrl;
  String? _authToken;
  WebSocketChannel? _channel;
  bool _isConnected = false;

  // LiveKit token received from room_state
  String? livekitToken;

  final _messageController = StreamController<WSMessage>.broadcast();
  Stream<WSMessage> get messages => _messageController.stream;

  final _clockOffsetController = StreamController<ClockSync>.broadcast();
  Stream<ClockSync> get clockSync => _clockOffsetController.stream;

  final _roomStateController = StreamController<Map<String, dynamic>>.broadcast();
  Stream<Map<String, dynamic>> get roomState => _roomStateController.stream;

  WebSocketService({this.baseUrl = 'ws://localhost:8081/ws'});

  bool get isConnected => _isConnected;

  /// Connect to WebSocket server with auth token
  Future<void> connect(String authToken) async {
    _authToken = authToken;
    final uri = Uri.parse('$baseUrl?token=$authToken');

    try {
      _channel = WebSocketChannel.connect(uri);
      _isConnected = true;

      _channel!.stream.listen(
        _handleMessage,
        onError: (error) {
          print('WebSocket error: $error');
          _isConnected = false;
        },
        onDone: () {
          print('WebSocket closed');
          _isConnected = false;
        },
      );

      print('WebSocket connected');
    } catch (e) {
      _isConnected = false;
      rethrow;
    }
  }

  /// Disconnect from WebSocket
  void disconnect() {
    _channel?.sink.close(status.goingAway);
    _channel = null;
    _isConnected = false;
  }

  /// Send a message to the server
  void send(WSMessageType type, {Map<String, dynamic>? payload}) {
    if (!_isConnected) {
      throw Exception('WebSocket not connected');
    }

    final message = {
      'type': _toWireType(type),
      if (payload != null) 'payload': payload,
    };

    _channel!.sink.add(jsonEncode(message));
  }

  /// Handle incoming WebSocket messages
  void _handleMessage(dynamic data) {
    try {
      final json = jsonDecode(data as String);
      final typeStr = json['type'] as String?;

      if (typeStr == null) return;

      // Handle clock sync separately
      if (typeStr == 'time_sync_request') {
        _handleTimeSyncRequest(json);
        return;
      } else if (typeStr == 'time_sync_result') {
        _handleTimeSyncResult(json);
        return;
      }

      // Handle room_state (extract LiveKit token)
      if (typeStr == 'room_state') {
        final payload = (json['payload'] as Map<String, dynamic>?) ?? json;
        if (payload['livekitToken'] != null) {
          livekitToken = payload['livekitToken'] as String;
        }
        _roomStateController.add(payload);
        return;
      }

      // Parse other message types
      final type = _parseMessageType(typeStr);
      if (type == null) return;

      final payload = (json['payload'] as Map<String, dynamic>?) ?? json;

      _messageController.add(WSMessage(
        type: type,
        payload: payload,
      ));
    } catch (e) {
      print('Error parsing WebSocket message: $e');
    }
  }

  /// Handle time sync request from server (NTP-style handshake)
  void _handleTimeSyncRequest(Map<String, dynamic> json) {
    final payload = (json['payload'] as Map<String, dynamic>?) ?? json;
    final t1 = payload['t0'] as int? ?? payload['t1'] as int? ?? 0;
    final t2 = DateTime.now().millisecondsSinceEpoch;
    final t3 = DateTime.now().millisecondsSinceEpoch;

    send(WSMessageType.timeSyncResponse, payload: {
      't0': t1,
      't1': t1,
      't2': t2,
      't3': t3,
    });
  }

  /// Handle time sync result from server
  void _handleTimeSyncResult(Map<String, dynamic> json) {
    final payload = (json['payload'] as Map<String, dynamic>?) ?? json;
    final offsetMs = (payload['offsetMs'] as num?)?.toInt() ?? 0;
    final rttMs = (payload['rttMs'] as num?)?.toInt() ?? 0;

    _clockOffsetController.add(ClockSync(
      offsetMs: offsetMs,
      rttMs: rttMs,
    ));
  }

  // ─── Host Control Messages ────────────────────────────────

  void sendPlay({required String trackId, int? positionMs}) {
    send(WSMessageType.play, payload: {
      'trackId': trackId,
      if (positionMs != null) 'positionMs': positionMs,
      'hostTimestamp': DateTime.now().millisecondsSinceEpoch,
    });
  }

  void sendPause({int? positionMs}) {
    send(WSMessageType.pause, payload: {
      if (positionMs != null) 'positionMs': positionMs,
      'hostTimestamp': DateTime.now().millisecondsSinceEpoch,
    });
  }

  void sendSeek({required int positionMs}) {
    send(WSMessageType.seek, payload: {
      'positionMs': positionMs,
      'hostTimestamp': DateTime.now().millisecondsSinceEpoch,
    });
  }

  void sendSkip({required String trackId}) {
    send(WSMessageType.skip, payload: {
      'trackId': trackId,
    });
  }

  void sendTrackChanged({
    required String trackId,
    required int trackIndex,
    required String title,
    required String artist,
    required int durationMs,
  }) {
    send(WSMessageType.trackChanged, payload: {
      'trackId': trackId,
      'trackIndex': trackIndex,
      'title': title,
      'artist': artist,
      'durationMs': durationMs,
    });
  }

  void sendStartMirrorMode() {
    send(WSMessageType.startMirrorMode);
  }

  void sendStopMirrorMode() {
    send(WSMessageType.stopMirrorMode);
  }

  void dispose() {
    disconnect();
    _messageController.close();
    _clockOffsetController.close();
    _roomStateController.close();
  }
}

class WSMessage {
  final WSMessageType type;
  final Map<String, dynamic> payload;

  WSMessage({required this.type, required this.payload});
}

class ClockSync {
  final int offsetMs;
  final int rttMs;

  ClockSync({required this.offsetMs, required this.rttMs});
}

File: web/src/services/websocketService.js
/**
 * WebSocket Service for TuneTogether Signaling Server
 * Handles real-time commands, clock sync, and room events
 */

const WS_BASE_URL = import.meta.env.VITE_WS_URL || 'ws://localhost:8081/ws';

class WebSocketService {
  constructor() {
    this.ws = null;
    this.connected = false;
    this.listeners = new Map();
    this.clockSyncListeners = [];
    this.driftListeners = [];
    this.roomStateListeners = [];
    this.reconnectAttempts = 0;
    this.maxReconnectAttempts = 5;
    this.reconnectDelay = 1000;
    this.livekitToken = null;
    this.authToken = null;
  }

  /**
   * Connect to WebSocket server with auth token
   */
  connect(authToken) {
    this.authToken = authToken;
    return new Promise((resolve, reject) => {
      const url = `${WS_BASE_URL}?token=${authToken}`;
      this.ws = new WebSocket(url);

      this.ws.onopen = () => {
        console.log('WebSocket connected');
        this.connected = true;
        this.reconnectAttempts = 0;
        resolve();
      };

      this.ws.onmessage = (event) => {
        this.handleMessage(event.data);
      };

      this.ws.onerror = (error) => {
        console.error('WebSocket error:', error);
        this.connected = false;
      };

      this.ws.onclose = () => {
        console.log('WebSocket closed');
        this.connected = false;
        this.attemptReconnect(authToken);
      };

      // Timeout after 10 seconds
      setTimeout(() => {
        if (!this.connected) {
          reject(new Error('WebSocket connection timeout'));
        }
      }, 10000);
    });
  }

  /**
   * Attempt to reconnect with exponential backoff
   */
  attemptReconnect(authToken) {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.error('Max reconnection attempts reached');
      return;
    }

    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);

    console.log(`Attempting reconnect in ${delay}ms (attempt ${this.reconnectAttempts})`);

    setTimeout(() => {
      this.connect(authToken).catch(console.error);
    }, delay);
  }

  /**
   * Disconnect from WebSocket
   */
  disconnect() {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
      this.connected = false;
    }
  }

  /**
   * Send a message to the server
   */
  send(type, payload = {}) {
    if (!this.connected || !this.ws) {
      console.error('WebSocket not connected');
      return;
    }

    const message = {
      type,
      ...payload,
    };

    this.ws.send(JSON.stringify(message));
  }

  /**
   * Handle incoming messages
   */
  handleMessage(data) {
    try {
      const message = JSON.parse(data);
      const { type } = message;

      // Handle clock sync
      if (type === 'timeSyncRequest') {
        this.handleTimeSyncRequest(message);
        return;
      } else if (type === 'timeSyncResult') {
        this.handleTimeSyncResult(message);
        return;
      }

      // Handle room state (contains LiveKit token)
      if (type === 'room_state') {
        const payload = message.payload || message;
        if (payload.livekitToken) {
          this.livekitToken = payload.livekitToken;
        }
        this.roomStateListeners.forEach((cb) => cb(payload));
        return;
      }

      // Handle drift correction
      if (type === 'driftCorrection' || type === 'drift_correction') {
        this.handleDriftCorrection(message);
        return;
      }

      // Notify listeners
      const listeners = this.listeners.get(type) || [];
      listeners.forEach((callback) => callback(message));
    } catch (error) {
      console.error('Error parsing WebSocket message:', error);
    }
  }

  /**
   * Handle time sync request (NTP-style)
   */
  handleTimeSyncRequest(message) {
    const t1 = message.t1; // Server sent time
    const t2 = Date.now(); // Client receive time
    const t3 = Date.now(); // Client send time (immediate)

    this.send('timeSyncResponse', {
      t1,
      t2,
      t3,
    });
  }

  /**
   * Handle time sync result
   */
  handleTimeSyncResult(message) {
    const { offsetMs, rttMs } = message;
    console.log(`Clock sync: offset=${offsetMs}ms, rtt=${rttMs}ms`);

    this.clockSyncListeners.forEach((callback) =>
      callback({ offsetMs, rttMs })
    );
  }

  /**
   * Phase 7: Handle drift correction command
   */
  handleDriftCorrection(message) {
    const { targetPositionMs, adjustmentRate } = message;
    console.log(`Drift correction: target=${targetPositionMs}ms, rate=${adjustmentRate}x`);

    this.driftListeners.forEach((callback) =>
      callback({ targetPositionMs, adjustmentRate })
    );
  }

  /**
   * Phase 7: Report playback position for drift monitoring
   */
  reportPlaybackPosition(positionMs, trackId) {
    this.send('playbackPositionReport', {
      positionMs,
      trackId,
      timestamp: Date.now(),
    });
  }

  /**
   * Subscribe to a message type
   */
  on(type, callback) {
    if (!this.listeners.has(type)) {
      this.listeners.set(type, []);
    }
    this.listeners.get(type).push(callback);

    // Return unsubscribe function
    return () => {
      const listeners = this.listeners.get(type) || [];
      const index = listeners.indexOf(callback);
      if (index > -1) {
        listeners.splice(index, 1);
      }
    };
  }

  /**
   * Subscribe to clock sync updates
   */
  onClockSync(callback) {
    this.clockSyncListeners.push(callback);
    return () => {
      const index = this.clockSyncListeners.indexOf(callback);
      if (index > -1) {
        this.clockSyncListeners.splice(index, 1);
      }
    };
  }

  /**
   * Phase 7: Subscribe to drift corrections
   */
  onDriftCorrection(callback) {
    this.driftListeners.push(callback);
    return () => {
      const index = this.driftListeners.indexOf(callback);
      if (index > -1) {
        this.driftListeners.splice(index, 1);
      }
    };
  }

  /**
   * Subscribe to room state updates (contains LiveKit token)
   */
  onRoomState(callback) {
    this.roomStateListeners.push(callback);
    return () => {
      const index = this.roomStateListeners.indexOf(callback);
      if (index > -1) this.roomStateListeners.splice(index, 1);
    };
  }

  /**
   * Get the LiveKit token received from room state
   */
  getLiveKitToken() {
    return this.livekitToken;
  }

  // ─── Host Control Messages ──────────────────────────────────

  sendPlay(trackId, positionMs = 0) {
    this.send('play', {
      payload: { trackId, positionMs, hostTimestamp: Date.now() },
    });
  }

  sendPause(positionMs = 0) {
    this.send('pause', {
      payload: { positionMs, hostTimestamp: Date.now() },
    });
  }

  sendSeek(positionMs) {
    this.send('seek', {
      payload: { positionMs, hostTimestamp: Date.now() },
    });
  }

  sendSkip(direction = 'next') {
    this.send('skip', {
      payload: { direction },
    });
  }

  sendTrackChanged(track) {
    this.send('track_changed', {
      payload: {
        trackId: track.id,
        trackIndex: track.orderIndex || 0,
        title: track.title,
        artist: track.artist,
        durationMs: track.durationMs,
      },
    });
  }
}

export default new WebSocketService();
